pkg/controller/util/planner/job_planner.go

```diff
@@ -0,0 +1,147 @@
+// Sorting is made by a hash to avoid assigning single-replica rs to the alphabetically smallest cluster.
+func (a ClustersByWeight) Less(i, j int) bool {
+	return (a[i].Weight > a[j].Weight) || (a[i].Weight == a[j].Weight && a[i].hash < a[j].hash)
+}
+
+func NewJobPlanner(preferences *fedschedulingv1a1.JobSchedulingPreference) *JobPlanner {
+	return &JobPlanner{
+		preferences: preferences,
+	}
+}
+
+// This assumes that there is atleast one available cluster and distributes
+// the parallelism or completions of the job equally among clusters or in
+// the ratio of weights if specified.
+func (p *JobPlanner) Plan(availableClusters []string, toDistribute int32, hashKey string) map[string]int32 {
+
+	named := func(name string, weight int32) *clusterWithWeight {
+		// Seems to work better than addler for our case.
+		hasher := fnv.New32()
+		hasher.Write([]byte(name))
+		hasher.Write([]byte(hashKey))
+
+		return &clusterWithWeight{
+			clusterName: name,
+			hash:        hasher.Sum32(),
+			Weight:      weight,
+		}
+	}
+
+	numClusters := len(availableClusters)
+	plan := make(map[string]int32, numClusters)
+	if numClusters == 0 {
+		return plan
+	}
+	// weightedList will only store those cluster values which have
+	// weights specified in the preferences.
+	weightedList := make([]*clusterWithWeight, 0, numClusters)
+	weightSum := int32(0)
+
+	for _, cluster := range availableClusters {
+		if weight, found := p.preferences.Spec.ClusterWeights[cluster]; found {
+			weightedList = append(weightedList, named(cluster, weight))
+			weightSum += weight
+		} else {
+			if weight, found := p.preferences.Spec.ClusterWeights["*"]; found {
+				weightedList = append(weightedList, named(cluster, weight))
+				weightSum += weight
+			} else {
+				// Clusters which do not have explicit weights given, while
+				// some others have, are treated equivalent to 0 weight.
+				// TODO(irfanurrehman): fix up a strategy about creating jobs
+				// with parallelism as 0. Right now we do not create jobs
+				// in those clusters which have 0 weightage.
+				//plan[cluster] = int32(0)
+			}
+		}
+	}
+	sort.Sort(ClustersByWeight(weightedList))
+
+	//distributeInThisLoop := toDistribute
+	remaining := toDistribute
+	if weightSum > 0 {
+		for _, cluster := range weightedList {
+			if remaining > 0 {
+				// Distribute the numbers as per the weight ratio, rounding fractions always up.
+				distributed := (toDistribute*cluster.Weight + weightSum - 1) / weightSum
+				distributed = minInt32(distributed, remaining)
+				plan[cluster.clusterName] = distributed
+				remaining -= distributed
+			}
+		}
+	} else {
+		// No weights, default distribution
+		distributionStep := toDistribute / int32(numClusters)
+		if distributionStep > 0 {
+			for _, clusterName := range availableClusters {
+				plan[clusterName] = distributionStep
+			}
+		}
+
+		quotient := toDistribute - (distributionStep * int32(numClusters))
```

Comment:
Isn't `quotient` just the modulus of `toDistribute` and the number of clusters? It should be an integer between zero and `numClusters-1`

Reply:
Thanks for pointing this out, I was wrongly using `quotient` for `remainder`. Updated!